### PR TITLE
Scalability testing framework 

### PR DESCRIPTION
**Pull request by amork**
Domovoi-based scalability testing framework.
Note: this is still work in progress. 
Pending items:
- automate deployment of the DynamoDB artifacts
- deployment of the aggregation Lambda
- stream/trigger configuration
- get rid of Locust

### CODE REVIEW

File: daemons/dss-scalability-test/.chalice/config.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "version": "2.0",
+  "app_name": "dss-scalability-test",
+  "stages": {
+    "dev": {
+      "api_gateway_stage": "api",
+      "environment_variables": {}
+    },
+    "roman": {
```
Comment (Bento007) [inline daemons/dss-scalability-test/.chalice/config.json]: We should probably make this a generic name like `user`.

File: daemons/dss-scalability-test/app.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+@app.dynamodb_stream_handler(table_name="scalability_test", batch_size=50)
+def handle_dynamodb_stream(event, context):
+    success_count = 0
+    failure_count = 0
+    records = 0
+    duration_sum = 0
+    run_id = None
+    for record in event['Records']:
+        if record['eventName'] == 'INSERT':
+            run_id = record['dynamodb']['NewImage']['run_id']['S']
```
Comment (Bento007) [inline daemons/dss-scalability-test/app.py]: is `run_id` supposed to change in every loop?

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
+hca
```
Comment (Bento007) [inline requirements.txt]: why is this here?

File: daemons/dss-scalability-test/app.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            'duration': current_time() - start_time - WAIT_CHECKOUT * 1000,
+            'success_count': success_count,
+            'fail_count': fail_count,
+            'created_on': datetime.datetime.now().isoformat(),
+            'expiration_ttl': expiration_ttl
+        }
+    )
+
+
+@app.sns_topic_subscriber("dss-scalability-test-run-" + os.environ["DSS_DEPLOYMENT_STAGE"])
```
Comment (kislyuk) [inline daemons/dss-scalability-test/app.py]: In the docs for the daemon, please list the SNS topics, dynamodb table, and their relationship to the state machine.

File: daemons/dss-scalability-test/app.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+            old_avg_duration = run_entry['Item']['average_duration']
+            old_count = run_entry['Item']['succeeded_count'] + run_entry['Item']['failed_count']
+
+            table.update_item(
+                Key=run_entry_pk,
+                UpdateExpression='SET '
+                                 'succeeded_count = succeeded_count + :success_count, '
+                                 'failed_count = failed_count + :failure_count, '
+                                 'executions = executions + :records, '
+                                 'average_duration = :new_average',
```
Comment (kislyuk) [inline daemons/dss-scalability-test/app.py]: It seems like you should use ADD instead of SET here.

File: daemons/dss-scalability-test/domovoilib/json_generator/__init__.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import json
```
Comment (kislyuk) [inline daemons/dss-scalability-test/domovoilib/json_generator/__init__.py]: Is this the real json generator for scaletest, or a stand-in until the code that @Bento007 is working on lands?

If it's the real thing, I would prefer it if it was broken out into its own PR, and probably moved to dss.util.

File: tests/scalability/scale_test_runner.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import uuid
+
+import argparse
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), 'sns')))  # noqa
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))  # noqa
+
+from sns import SnsClient
+
+UNIT_OF_TIME = 1 * 1000 * 1000  # 1 sec in microseconds
```
Comment (kislyuk) [inline tests/scalability/scale_test_runner.py]: Constants should be called something more descriptive. Perhaps `ONE_SECOND_IN_MS`

File: tests/scalability/sns/__init__.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import time
```
Comment (kislyuk) [inline tests/scalability/sns/__init__.py]: Can you please try replacing these two files with a higher level Python threading API? I suggest using a simple concurrent.futures.map or threadpool context manager. It should make this code much simpler.

File: daemons/dss-scalability-test/domovoilib/json_generator/__init__.py (hunk 6 above)
Comment (Bento007) [reply daemons/dss-scalability-test/domovoilib/json_generator/__init__.py]: it's a standin

File: daemons/dss-checkout-sfn/app.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def schedule_copy(event, context):
         scheduled += 1
     return {"files_scheduled": scheduled,
             "dst_location": get_dst_bundle_prefix(bundle_fqid, version),
-            "wait_time_seconds": 10}
+            "wait_time_seconds": 30}
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-checkout-sfn/app.py]: I agree, I'm going to re-organize checkout service code in a separate PR

File: daemons/dss-scalability-test/.chalice/config.json [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{
+  "version": "2.0",
+  "app_name": "dss-scalability-test",
+  "stages": {
+    "dev": {
+      "api_gateway_stage": "api",
+      "environment_variables": {
+        "HOME": "/tmp"
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-scalability-test/.chalice/config.json]: it is still necessary for creating tempfiles and temp directories, adding pseudo-comment

File: daemons/dss-scalability-test/app.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), 'domovoilib'))  # noqa
+sys.path.insert(0, pkg_root)  # noqa
+
+from dss import stepfunctions
+from dss.stepfunctions import generator
+
+from hca.dss import DSSClient
+from json_generator import generate_sample
+
+AWS_MIN_CHUNK_SIZE = 64 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-scalability-test/app.py]: Done

File: daemons/dss-scalability-test/app.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
+import datetime
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-scalability-test/app.py]: done

File: daemons/dss-scalability-test/app.py [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+from dss import stepfunctions
+from dss.stepfunctions import generator
+
+from hca.dss import DSSClient
+from json_generator import generate_sample
+
+AWS_MIN_CHUNK_SIZE = 64 * 1024 * 1024
+WAIT_CHECKOUT = 3
+PARALLELIZATION_FACTOR = 10
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-scalability-test/app.py]: done

File: daemons/dss-scalability-test/app.py [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+
+
+def current_time():
+    return int(round(time.time() * 1000))
+
+
+def upload_bundle(event, context, branch_id):
+    app.log.info("Upload bundle")
+    with tempfile.TemporaryDirectory() as src_dir:
+        # TODO: insert json generator here
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-scalability-test/app.py]: Updated TODO, Trent is waiting for this PR to merge before he can plugin his json generator

File: daemons/dss-scalability-test/app.py (hunk 4 above)
Comment (amork) [reply daemons/dss-scalability-test/app.py]: Done

File: tests/scalability/scale_test_runner.py (hunk 7 above)
Comment (amork) [reply tests/scalability/scale_test_runner.py]: done

**Comment by kislyuk** [comment]
@amork is this ready for re-review?

File: tests/scalability/sns/__init__.py (hunk 8 above)
Comment (amork) [reply tests/scalability/sns/__init__.py]: Done

File: daemons/dss-scalability-test/app.py (hunk 2 above)
Comment (amork) [reply daemons/dss-scalability-test/app.py]: no, updated

File: daemons/dss-scalability-test/README.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
+# HCA DSS: Scalability Testing framework
+
+This is a scalability test framework for the replicated data storage system (aka the "blue box") of
+the [Human Cell Atlas](https://www.humancellatlas.org/). 
+
+#### About the scalability testing framework
+
+1. The scalability test frameworks is based on AWS Step Function. Workflow definition resembles smoke test for DSS.
```
Comment (kislyuk) [inline daemons/dss-scalability-test/README.md]: s/frameworks/framework
s/Function/Functions

File: daemons/dss-scalability-test/README.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+2. The execution is triggered by sending SNS messages to dss-scalability-test-run-[STAGE} topic
```
Comment (kislyuk) [inline daemons/dss-scalability-test/README.md]: Please fix the brackets

File: daemons/dss-scalability-test/README.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+3. The scalability test writes results of execution of individual executions and aggregated run metrics into the 
+following DynamoDB tables: scalability_test_run, scalability_test
+4. The SFN execution is initiatied and starts by entering WAIT step. Wait is configured to end at the 5 minute intervals 
+to accomodate the AWS limit on starting SFN executions and enable generation of bursts of load
+5.  Once all parallel branches of execution are done, it writes summary of the run in DynamoDB
+6. DynamoDB is configured to stream new records into Lambda which aggregates the results and writes incremental metrics
+back into SynamoDB
```
Comment (kislyuk) [inline daemons/dss-scalability-test/README.md]: s/SynamoDB/DynamoDB

File: daemons/dss-scalability-test/README.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+7. CloudWatch dashboard has been configured to display relevant execution metrics
```
Comment (kislyuk) [inline daemons/dss-scalability-test/README.md]: This sounds like one-off manual account configuration. Please replace this line with instructions for how to use the dashboard in general for this purpose, or a specific recipe.

File: daemons/dss-scalability-test/domovoilib/dynamodb/__init__.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
+import boto3
+
+dynamodb_client = boto3.client('dynamodb')
+
+SCALABILITY_TEST_TABLE = 'scalability_test'
+SCALABILITY_TEST_RUN_TABLE = 'scalability_test_run'
```
Comment (kislyuk) [inline daemons/dss-scalability-test/domovoilib/dynamodb/__init__.py]: These table names are not sufficiently differentiated. Can you rename scalability_test to something more specific (scalability_test_results?)

File: daemons/dss-scalability-test/requirements.txt [hunk 20]
```diff
@@ -0,0 +1,49 @@
+hca >= 3.2.0
```
Comment (kislyuk) [inline daemons/dss-scalability-test/requirements.txt]: Why is this requirements file not a link to the one in the parent directory? If it's a superset of the one in the parent directory, then let's make it include that one with "-r ../../requirements.txt".

File: iam/policy-templates/dss-scalability-test-lambda.json [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      ],
+      "Resource": "arn:aws:logs:*:*:*"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:List*",
+        "s3:Get*",
+        "s3:PutObject*",
+        "s3:DeleteObject*"
```
Comment (kislyuk) [inline iam/policy-templates/dss-scalability-test-lambda.json]: This is not an appropriate permission to give to this daemon. Let's review this offline.

File: requirements-dev.txt [hunk 22]
```diff
@@ -25,17 +26,19 @@ cryptography==2.1.3
 docker==2.6.1
 docker-pycreds==0.2.1
 docutils==0.14
-domovoi==1.4.5
+domovoi==1.5.1
```
Comment (kislyuk) [inline requirements-dev.txt]: Please rebase this on master - I committed a change that conflicts with this.

File: daemons/dss-scalability-test/README.md (hunk 15 above)
Comment (amork) [reply daemons/dss-scalability-test/README.md]: done

File: daemons/dss-scalability-test/README.md (hunk 16 above)
Comment (amork) [reply daemons/dss-scalability-test/README.md]: done

File: daemons/dss-scalability-test/README.md (hunk 17 above)
Comment (amork) [reply daemons/dss-scalability-test/README.md]: done

File: daemons/dss-scalability-test/app.py (hunk 5 above)
Comment (amork) [reply daemons/dss-scalability-test/app.py]: In this case the attributes already exists so using SET is more appropriate.

File: daemons/dss-scalability-test/requirements.txt (hunk 20 above)
Comment (amork) [reply daemons/dss-scalability-test/requirements.txt]: agreed, done

File: iam/policy-templates/dss-scalability-test-lambda.json (hunk 21 above)
Comment (amork) [reply iam/policy-templates/dss-scalability-test-lambda.json]: Delete was no longer needed, removed

File: requirements-dev.txt (hunk 22 above)
Comment (amork) [reply requirements-dev.txt]: done
